habitat_baselines/il/trainers/pacman_trainer.py

```diff
@@ -0,0 +1,660 @@
+    used in EmbodiedQA (Das et. al.;CVPR 2018)
+    Paper: https://embodiedqa.org/paper.pdf.
+    """
+    supported_tasks = ["EQA-v0"]
+
+    def __init__(self, config=None):
+        super().__init__(config)
+
+        self.device = (
+            torch.device("cuda", self.config.TORCH_GPU_ID)
+            if torch.cuda.is_available()
+            else torch.device("cpu")
+        )
+
+        if config is not None:
+            logger.info(f"config: {config}")
+
+    def _save_nav_results(
+        self,
+        ckpt_path: int,
+        ep_id: int,
+        questions: torch.Tensor,
+        imgs: List[np.ndarray],
+        q_vocab_dict: Dict,
+        results_dir: str,
+        writer: TensorboardWriter,
+        video_option: list,
+    ) -> None:
+
+        r"""For saving NAV-PACMAN eval results.
+        Args:
+            ckpt_path: path of checkpoint being evaluated
+            ep_id: episode id (batch index)
+            questions: input question to model
+            imgs: images' tensor containing input frames
+            q_vocab_dict: question vocab dictionary
+            results_dir: dir to save results
+            writer: tensorboard writer
+            video_option: ["disk", "tb"]
+        Returns:
+            None
+        """
+
+        question = questions[0]
+
+        ckpt_epoch = ckpt_path[ckpt_path.rfind("/") + 1 : -5]
+        results_dir = os.path.join(results_dir, ckpt_epoch)
+        ckpt_no = ckpt_epoch[6:]
+
+        q_string = q_vocab_dict.token_idx_2_string(question)
+        frames_with_text = []
+        for frame in imgs:
+            border_width = 32
+            font = cv2.FONT_HERSHEY_SIMPLEX
+            color = (0, 0, 0)
+            scale = 0.3
+            thickness = 1
+
+            frame = cv2.copyMakeBorder(
+                frame,
+                border_width,
+                border_width,
+                border_width,
+                border_width,
+                cv2.BORDER_CONSTANT,
+                value=(255, 255, 255),
+            )
+
+            frame = cv2.putText(
+                frame,
+                "Question: " + q_string,
+                (10, 15),
+                font,
+                scale,
+                color,
+                thickness,
+            )
+
+            frames_with_text.append(frame)
+        generate_video(
+            video_option,
+            results_dir,
+            frames_with_text,
+            ep_id,
+            ckpt_no,
+            {},
+            writer,
+            fps=5,
+        )
+
+    def train(self) -> None:
+        r"""Main method for training Navigation model of EQA.
+
+        Returns:
+            None
+        """
+        config = self.config
+
+        with habitat.Env(config.TASK_CONFIG) as env:
+            nav_dataset = (
+                NavDataset(
+                    config,
+                    env,
+                    self.device,
+                )
+                .shuffle(1000)
```

Comment:
Magic constant has to be constant with name or seed from the config.

Reply:
@mathfac The `shuffle` parameter acts as a buffer for shuffling the dataset. ([Source](https://webdataset.readthedocs.io/en/stable/#using-webdataset))

> shuffle(n): shuffle the dataset with a buffer of size n

Should it be added to the config?